pkg/devserver/lifecycle.go

```diff
@@ -45,3 +53,60 @@ func (l lifecycle) OnFunctionScheduled(
 		Cron:         event.Cron,
 	})
 }
+
+func (l lifecycle) OnFunctionFinished(
+	ctx context.Context,
+	id state.Identifier,
+	item queue.Item,
+	resp state.DriverResponse,
+	s state.State,
+) {
+	now := time.Now()
+	data := map[string]interface{}{
+		"function_id": s.Function().Slug,
+		"run_id":      id.RunID.String(),
+	}
+
+	origEvt := s.Event()
+
+	if dataMap, ok := origEvt["data"].(map[string]interface{}); ok {
+		if inngestObj, ok := dataMap[consts.InngestEventDataPrefix].(map[string]interface{}); ok {
+
+			if dataValue, ok := inngestObj[consts.InvokeCorrelationId].(string); ok {
+				logger.From(ctx).Debug().Str("data_value_str", dataValue).Msg("data_value")
+				data[consts.InvokeCorrelationId] = dataValue
+			}
+		}
+	}
+
+	if resp.Err != nil {
+		data["error"] = resp.UserError()
+	} else {
+		data["result"] = resp.Output
+	}
+
+	evt := event.Event{
+		ID:        ulid.MustNew(uint64(now.UnixMilli()), rand.Reader).String(),
+		Name:      event.FnCompletedName,
+		Timestamp: now.UnixMilli(),
+		Data:      data,
+	}
+
+	logger.From(ctx).Debug().Interface("event", evt).Msg("function completed event")
+
+	byt, err := json.Marshal(evt)
+	if err != nil {
+		logger.From(ctx).Error().Err(err).Msg("error marshalling function completion event")
+		return
+	}
+
+	_ = l.pb.Publish(
+		ctx,
+		l.eventTopic,
+		pubsub.Message{
+			Name:      event.EventReceivedName,
+			Data:      string(byt),
+			Timestamp: now,
+		},
+	)
```

Comment:
Sending this is pretty critical, though lifecycle methods probably shouldn't contain crticial flows. Thinking this should move outside of a lifecycle method.

Reply:
We definitely need to log a hard error here.  Realistically, the one component that _cannot_ go down are our message streams.  Without those, things are in dire straights (as they are for every large company).

Reply:
Yeah. Let's keep the lifecycle contract the same and I'll pull it out into another function.